OpenTracing Bridge: allow more generic carriers

bboreham opened the pull request:
Instead of insisting that the carrier is an `HTTPHeaders`, cast it or adapt it to the interface we need - `TextMapCarrier`.

~It's a bit long-winded to wrap the read and write side separately, but this gives maximum flexibility.~

Fixes #2137 - @kvrhdn has used this fork to create a working program.

kvrhdn commented:
I can confirm this fixes our issue, propagation context can be injected and extracted without error now 🙂

Aneurysm9 commented on bridge/opentracing/bridge.go:
```diff
@@ -620,6 +620,25 @@ func (s fakeSpan) SpanContext() trace.SpanContext {
+// adapt OpenTracing TextMapWriter to OpenTelemetry propagation.TextMapCarrier
+type textMapWriterAdapter struct {
+	t ot.TextMapWriter
+}
```
```suggestion
}

var _ propagation.TextMapCarrier = textMapWriterAdapter{}
var _ propagation.TextMapCarrier = textMapReaderAdapter{}
```

I'd like to see tests, but at the least we should assert that the interfaces are satisfied.

bboreham replied:
The effect of these assertions is already covered by the assignments to `tmCarrier` which is of type `propagation.TextMapCarrier`.

MrAlias commented on bridge/opentracing/bridge.go:
```diff
@@ -620,6 +620,25 @@ func (s fakeSpan) SpanContext() trace.SpanContext {
+
 // Inject is a part of the implementation of the OpenTracing Tracer
 // interface.
 //
```
Docs need to be updated that they support more than HTTP.

bboreham replied:
Is this a reference to the line below?  The `format` parameter is still checked to be `ot.HTTPHeaders`, here:
https://github.com/open-telemetry/opentelemetry-go/pull/2141/files#diff-0f0430a4013bb87f6c4d12fa2f8a997e871e952c5a3f7f279aff88b54eb1cf32R674-R676

MrAlias commented on bridge/opentracing/bridge.go:
```diff
@@ -620,6 +620,25 @@ func (s fakeSpan) SpanContext() trace.SpanContext {
 	return s.sc
 }
 
+// adapt OpenTracing TextMapWriter to OpenTelemetry propagation.TextMapCarrier
```
```suggestion
// textMapWriterAdapter adapts an OpenTracing TextMapWriter to an OpenTelemetry propagation.TextMapCarrier.
```

MrAlias commented on bridge/opentracing/bridge.go:
```diff
@@ -635,21 +654,54 @@ func (t *BridgeTracer) Inject(sm ot.SpanContext, format interface{}, carrier int
 	return nil
 }
 
+// adapt OpenTracing TextMapReader to OpenTelemetry propagation.TextMapCarrier
```
```suggestion
// textMapReaderAdapter adapts an OpenTracing TextMapReader to an OpenTelemetry propagation.TextMapCarrier.
```

MrAlias commented on bridge/opentracing/bridge.go:
```diff
@@ -620,6 +620,25 @@ func (s fakeSpan) SpanContext() trace.SpanContext {
+}
+
+func (a textMapWriterAdapter) Keys() []string {
+	// not implemented
```
Are both the `textMap(Writer|Reader)Adapter` needed? It seems like they could be unified into a single type that would wrap both a OT propagation types.

bboreham replied:
How were you thinking to do that?

MrAlias replied:
I would merge the two by including two fields for both types of OT carriers removing the unimplemented methods and replacing with the equivalent implemented ones. You would only add checks to the methods to see if the respective carrier field was set before preforming any actions.

bboreham replied:
OK I gave it a go.

MrAlias reviewed (changes requested):
Thanks for the contribution, it is definitely welcomed as we ultimately want to support more than just HTTP carriers for the OT bridge.

These changes need to include tests. Primarily the tests need to ensure behavior of the introduced types are preserved.

MrAlias commented on bridge/opentracing/bridge.go:
```diff
@@ -620,6 +620,45 @@ func (s fakeSpan) SpanContext() trace.SpanContext {
 	return s.sc
 }
 
+// adapt OpenTracing TextMapReader or TextMapWriter to OpenTelemetry propagation.TextMapCarrier
```
```suggestion
// textMapAdapter adapts an OpenTracing TextMapReader or TextMapWriter to an OpenTelemetry propagation.TextMapCarrier.
```

hickeyma commented:
Ping @bboreham. Rebase needed and some feedback to look at.

bboreham commented:
Ack.  It hasn't been high enough up my priority list lately.

bboreham commented:
Rebased and added unit tests for the Inject/Extract functionality added in this PR.

Note that `ot.HTTPHeadersCarrier` needs special treatment - `Set()` modifies the keys to meet HTTP header expectations - so we can't use the generic wrapper I added in that case.

It appears baggage was never propagated by `BridgeTracer`, so that is not included in the test.

Aneurysm9 commented:
Does https://github.com/open-telemetry/opentelemetry-go/pull/2911 meet the needs here, or is there more required?

bboreham commented:
> Does https://github.com/open-telemetry/opentelemetry-go/pull/2911 meet the needs here

Strictly no, however I think #2911 makes it easier to work around the deficiency.
Perhaps you will update the [doc](https://github.com/opentracing/opentracing-go/blob/10b1cf09e00bdc84234b8c7a4b4d4e4afe64de87/propagation.go#L71) says it should work with `TextMapWriter`?

Do you have any response to https://github.com/open-telemetry/opentelemetry-go/pull/2141#issuecomment-1101314799 ?

zalegrala commented:
@MrAlias Do you have any further feedback for this PR please?

MrAlias commented on bridge/opentracing/bridge.go:
```diff
@@ -635,18 +674,22 @@ func (t *BridgeTracer) Inject(sm ot.SpanContext, format interface{}, carrier int
 		return ot.ErrUnsupportedFormat
 	}
-	hhcarrier, ok := carrier.(ot.HTTPHeadersCarrier)
+	// If carrier implements the required interface directly, use that
```
```suggestion
	// If carrier implements the required interface directly, use that.
```

MrAlias commented on bridge/opentracing/bridge.go:
```diff
@@ -635,18 +674,22 @@ func (t *BridgeTracer) Inject(sm ot.SpanContext, format interface{}, carrier int
+	tmCarrier, ok := carrier.(propagation.TextMapCarrier)
 	if !ok {
-		return ot.ErrInvalidCarrier
+		tmWriter, ok := carrier.(ot.TextMapWriter) // otherwise see if we can wrap it
```
```suggestion
		tmWriter, ok := carrier.(ot.TextMapWriter) // Otherwise, see if we can wrap it.
```

dmathieu commented:
Closing this as stale.
This is not to say the proposal cannot be accepted, but the PR had change suggestions that were not applied in 2 years. And getting this ready to ship now would also require fixing several conflicts.
